### PR TITLE
WIP: Scale ha combined

### PR DESCRIPTION
**Pull request by jayunit100**
This is a WIP for HA implementation from the HA proposal (here)[proposals/high-availability.md].  Just posting to share, definitely not ready to merge, lots of debug floating around, but open to any comments or ideas.

### CODE REVIEW

**Comment by jayunit100** [comment]
@timothysc @rrati  fyi , PR is here for tracking (obviously tons of rebasing and I want to remove some debugging from lease.go once we test at scale)

**Comment by timothysc** [comment]
/cc @davidopp @bgrant0607 @wojtek-t 
Still a WIP and needs experimental, but comments solicited.  We can gather them while we fix up the other items.

**Comment by davidopp** [comment]
@mikedanese would you be interested in taking a look at this?

File: pkg/api/types.go [hunk 1]
```diff
@@ -2154,3 +2154,32 @@ type RangeAllocation struct {
...
+	Spec LockSpec `json:"spec,omitempty"`
+}
+
+// LockList is a list of all locks in a namespace.
+type LockList struct {
+	TypeMeta `json:",inline"`
+	ListMeta `json:"metadata,omitempty"`
+
+	Items []Lock `json:"locks"`
+}
```
Comment (mikedanese) [inline pkg/api/types.go]: @bgrant0607 for API changes.

File: pkg/tools/ha/lease.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//This utility uses the Lock API in the client to acquire a lock.
+//It provides a uniform and consistent semantics for.
+// - Logging how many times a lease is gained/lost.
+// - Starting/Stopping a daemon (via callbacks).
+// - Logging errors which might occur if a daemon's lease isn't consistent with its running state.
+package ha
```
Comment (mikedanese) [inline pkg/tools/ha/lease.go]: pkg/lease or pkg/util/lease would be a better place for this.

File: pkg/tools/ha/lease.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+	LeaseLost     func(lu *LeaseUser) bool
+	Cli           *client.Client
+	LeaseUserInfo *LeaseUser
+}
+
+// RunHA runs a process in a highly available fashion.
+func RunHA(kubecfg string, master string, start func(l *LeaseUser) bool, stop func(l *LeaseUser) bool, lockName string) {
+
+	//We need a kubeconfig in order to use the locking API, so we create it here.
+	kubeconfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
```
Comment (mikedanese) [inline pkg/tools/ha/lease.go]: This is not the right place for config generation. A kubeClient should be passed in.

File: pkg/tools/ha/lease.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+	// They should also update information in the LeaseUserInfo struct
+	// about the state of the lease owner.
+	LeaseGained   func(lu *LeaseUser) bool
+	LeaseLost     func(lu *LeaseUser) bool
+	Cli           *client.Client
+	LeaseUserInfo *LeaseUser
+}
+
+// RunHA runs a process in a highly available fashion.
+func RunHA(kubecfg string, master string, start func(l *LeaseUser) bool, stop func(l *LeaseUser) bool, lockName string) {
```
Comment (mikedanese) [inline pkg/tools/ha/lease.go]: This name is very descriptive. HA tells me nothing.

File: cmd/kube-controller-manager/controller-manager.go [hunk 5]
```diff
@@ -46,10 +47,35 @@ func main() {
...
+		//Even though we're exiting, we should set this flag before dying just for completeness.
+		leaseUserInfo.Running = false
+		os.Exit(0)
+
+		//this should never be reached
+		return true
+	}
+
+	//This starts a thread that continues running.
+	ha.RunHA(s.Kubeconfig, s.Master, startRC, endRC, "ha.cm.lock")
```
Comment (mikedanese) [inline cmd/kube-controller-manager/controller-manager.go]: I really dislike that we are doing all this in init. We should move it to a main.

File: cmd/kube-controller-manager/controller-manager.go [hunk 6]
```diff
@@ -46,10 +47,35 @@ func main() {
...
+		leaseUserInfo.Running = true
+		if err := s.Run(pflag.CommandLine.Args()); err != nil {
+			fmt.Fprintf(os.Stderr, "%v\n", err)
+			glog.Infof("EXITING NOW ! Killed")
+			os.Exit(1)
+		}
+		return true
+	}
+
+	endRC := func(leaseUserInfo *ha.LeaseUser) bool {
```
Comment (mikedanese) [inline cmd/kube-controller-manager/controller-manager.go]: stopRC? Also this is the controller-manager which runs five+ controllers. What does RC mean here?

File: cmd/kube-controller-manager/controller-manager.go [hunk 7]
```diff
@@ -46,10 +47,35 @@ func main() {
...
+	}
+
+	//This starts a thread that continues running.
+	ha.RunHA(s.Kubeconfig, s.Master, startRC, endRC, "ha.cm.lock")
 
-	if err := s.Run(pflag.CommandLine.Args()); err != nil {
-		fmt.Fprintf(os.Stderr, "%v\n", err)
-		os.Exit(1)
+	for true {
+		glog.Infof("CM lease loop is running...")
```
Comment (mikedanese) [inline cmd/kube-controller-manager/controller-manager.go]: Is this for debugging? verflag.PrintAndExitIfRequested() is unreachable now.

File: pkg/api/types.go [hunk 8]
```diff
@@ -2154,3 +2154,32 @@ type RangeAllocation struct {
...
+	// How long the lock is valid.
+	LeaseTime uint64 `json:"duration"`
+	// When the lock was first acquired.
+	AcquiredTime string `json:"atime,omitempty"`
+	// When the lock was last renewed.
+	RenewTime string `json:"rtime,omitempty"`
+}
+
+// Lock represents the configuration of a resource lock.
+type Lock struct {
```
Comment (mikedanese) [inline pkg/api/types.go]: Can we call this lease instead of lock?

File: pkg/api/types.go [hunk 9]
```diff
@@ -2154,3 +2154,32 @@ type RangeAllocation struct {
...
+// LockSpec is the specification of a resource lock.
+type LockSpec struct {
+	// Who currently has the lock.
+	HeldBy string `json:"heldby"`
+	// How long the lock is valid.
+	LeaseTime uint64 `json:"duration"`
+	// When the lock was first acquired.
+	AcquiredTime string `json:"atime,omitempty"`
+	// When the lock was last renewed.
+	RenewTime string `json:"rtime,omitempty"`
```
Comment (mikedanese) [inline pkg/api/types.go]: json names should match the field names but be lowerCamelCase

**Comment by mikedanese** [comment]
Please rename the PR to be something more descriptive.

**Comment by mikedanese** [comment]
This seems to have quite a bit of overlap with #12324. @karlkfi

File: pkg/api/types.go (hunk 8 above)
Comment (mikedanese) [inline pkg/api/types.go]: I take it back. I think Lock or Lease is a good name.

**Comment by mikedanese** [comment]
@jayunit100 I think we should divide this into multiple PRs that build on eachother. A logical division would be:

1. Lock/Lease API and resource in the experimental API (maybe include pkg/client and kubectl integration but that could also be separate)
2. Refactor podmaster.go into a library that lives in pkg/lease that uses the Lock/Lease API rather than etcd to implement useful patterns.
3. Retrofit controller-manager and scheduler to use pkg/lease
4. ...

This will make this chunk of work much easier to review and hopefully we can merge faster. WDYT?

File: cmd/kube-controller-manager/controller-manager.go (hunk 7 above)
Comment (jayunit100) [inline cmd/kube-controller-manager/controller-manager.go]: well, at this point, the daemon needs to block.  probably i can move the verflag up...

**Comment by jayunit100** [comment]
Hi mike.  I'm okay with that, i'll let @timothysc and @rrati chime in with opinions.  my initial thoughts are , 

- if we want to split, split into maybe just 2 PRs : One for lock+lease, and a second for scheduler, and controller.  
- Im happy to also refactor Podmaster as a separate task, but its not dependent on this thread of work.

lets see what the other guys say and then decide if we want to split (the lazy part of me feels that at this point it might be easier not to split) :)... but i see your point.

File: cmd/kube-controller-manager/controller-manager.go (hunk 6 above)
Comment (jayunit100) [inline cmd/kube-controller-manager/controller-manager.go]: yup this is supposed to be CM.  i thought i fixed that :)...

**Comment by jayunit100** [comment]
MOVED HERE https://github.com/kubernetes/kubernetes/pull/12884 .

Sorry for the confusion folks.

File: pkg/api/types.go (hunk 9 above)
Comment (rrati) [inline pkg/api/types.go]: I noticed some json names that weren't the same in the api and I thought duration was easier to deal with on the json side then the longer versions.  I have no problems changing though.

**Pull request closed without merging**